# Introduction

- Machine Learning is making the computer learn from studying data and statistics.

* Machine Learning is a step into the direction of artificial intelligence (AI).

+ Machine Learning is a program that analyses data and learns to predict the outcome.

### Overview
In this lesson, I will be:
- going back to mathematics and study statistics, and see how to calculate important numbers based on data sets.

+ learning how to use various Python modules to get the answers I need.

* learning how to make functions that are able to predict the outcome based on what I've learned.

# Data Set
In the mind of a computer, a **data set** is any collection of data. It can be anything from an array to a complete database. In Machine Learning it is common to work with very large data sets.

In [2]:
# Example of an array:

[99,86,87,88,111,86,103,87,94,78,77,85,86]

[99, 86, 87, 88, 111, 86, 103, 87, 94, 78, 77, 85, 86]

Example of a database:

|Carname	| Color |	Age |	Speed |	AutoPass|
|    :---:  |  :---:  |  :---:  | :---: | :---: |
|BMW |	red	 | 5	| 99	| Y|
|Volvo	| black |	7 |	86 |	Y|
|VW	| gray |	8 |	87 |	N|
|VW |	white |	7 |	88 |	Y|
|Ford |	white |	2 |	111 |	Y|
|VW |	white |	17 |	86 |	Y|
|Tesla |	red |	2 |	103 |	Y|
|BMW |	black |	9	| 87	| Y|
|Volvo |	gray |	4 |	94|	 N|
|Ford |	white |	11	| 78	| N|
|Toyota|	gray	| 12	| 77 |	N|
|VW | white| 9 |	85	| N|
|Toyota |	blue |	6	| 86	| Y|


By looking at the array, we can guess that the average value is probably around 80 or 90, and we are also able to determine the highest value and the lowest value, but what else can we do?

And by looking at the database we can see that the most popular color is white, and the oldest car is 17 years, but what if we could predict if a car had an AutoPass, just by looking at the other values?

That is what Machine Learning is for! Analyzing data and predicting the outcome!

# Data Types
To analyze data, it is important to know what type of data we are dealing with.

We can split the data types into three main categories:

1. Numerical
2. Categorical
3. Ordinal


1. Numerical data are numbers, and can be split into two numerical categories:

    * Discrete Data: numbers that are limited to integers. Example: The number of cars passing by.
    - Continuous Data: numbers that are of infinite value. Example: The price of an item, or the size of an item


2. Categorical data are values that cannot be measured up against each other. Example: a color value, or any yes/no values.

3. Ordinal data are like categorical data, but can be measured up against each other. Example: school grades where A is better than B and so on.

By knowing the data type of your data source, you will be able to know what technique to use when analyzing them.

More on statistics and analyzing data as I go on.

# Mean, Median, and Mode
What can we learn from looking at a group of numbers?

In Machine Learning (and in mathematics) there are often three values that interests us:

* Mean - The average value
- Median - The mid point value
+ Mode - The most common value

Example: We have registered the speed of 13 cars:

**speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]**

What is the average, the middle, or the most common speed value?

In [3]:
speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

### Mean
The mean value is the average value.

To calculate the mean, find the sum of all values, and divide the sum by the number of values:

In [4]:
# (99+86+87+88+111+86+103+87+94+78+77+85+86) / 13 = 89.77
total = 99+86+87+88+111+86+103+87+94+78+77+85+86
mean = total/13
print(mean)

89.76923076923077


The above can simply be solved using the python library NumPy:

In [6]:
import numpy as np

mean_speed = np.mean(speed)

print(mean_speed)

89.76923076923077


### Median
The median value is the value in the middle, after you have sorted all the values:

77, 78, 85, 86, 86, 86, 87, 87, 88, 94, 99, 103, 111

It is important that the numbers are sorted before you can find the median.

The NumPy module has a method for this:

In [62]:
import numpy

speed = [99,86,87,88,111,86,103,87,94,78,77,85,86]

x = numpy.median(speed)

print(x, "\n")

# or but this second step is not necessary

sort_speed = np.sort(speed)
print(sort_speed, "\n")

median_speed = np.median(speed)
print(median_speed)
print()

# median_sort_speed = np.sort(np.median(speed)) AxisError: axis -1 is out of bounds for array of dimension 0

87.0 

[ 77  78  85  86  86  86  87  87  88  94  99 103 111] 

87.0



Remember when calculating the median, if there are two numbers in the middle, divide the sum of those numbers by two.

speed = [77, 78, 85, 86, 86, 86, 87, 87, 94, 98, 99, 103]

middle_number = (86 + 87) / 2

### Mode
The Mode value is the value that appears the most number of times:

99, 86, 87, 88, 111, 86, 103, 87, 94, 78, 77, 85, 86 = 86

The **SciPy module** has a method for this:

In [68]:
from scipy import stats

x = stats.mode(speed)
print(x)

ModeResult(mode=array([86]), count=array([3]))


# Standard Deviation

#### What is Standard Deviation?
Standard deviation is a number that describes how spread out the values are.

**A low standard deviation means that most of the numbers are close to the mean (average) value.**

**A high standard deviation means that the values are spread out over a wider range.**

Example: This time we have registered the speed of 7 cars:

speed = [86,87,88,86,87,85,86]

The standard deviation is:

0.9

Meaning that **most of the values are within the range of 0.9 from the mean value, which is 86.4.**

Let us do the same with a selection of numbers with a wider range:

speed = [32,111,138,28,59,77,97]

The standard deviation is:

37.85

Meaning that **most of the values are within the range of 37.85 from the mean value, which is 77.4.**

As you can see, a higher standard deviation indicates that the values are spread out over a wider range.

The NumPy module has a method to calculate the standard deviation:



In [40]:
low_std_speed = [86,87,88,86,87,85,86]

x = np.std(low_std_speed)
print(x)
print()

high_std_speed = [32,111,138,28,59,77,97]

y = np.std(high_std_speed)
print(y)

0.9035079029052513

37.84501153334721


# Variance
Variance is another number that indicates how spread out the values are.

In fact, if you take the **square root of the variance,** you get the standard deviation!

Or the other way around, if you **multiply the standard deviation by itself,** you get the variance!

To calculate the variance you have to do as follows:

1. Find the mean:

In [42]:
mean = (32+111+138+28+59+77+97) / 7
print(mean)

77.42857142857143


2. For each value, find the difference from the mean:

` 32 - 77.4 = -45.4
111 - 77.4 =  33.6
138 - 77.4 =  60.6
 28 - 77.4 = -49.4
 59 - 77.4 = -18.4
 77 - 77.4 = - 0.4
 97 - 77.4 =  19.6`

3. For each difference, find the square value:

$(-45.4)^{2}$ = 2061.16

$(33.6)^{2}$ = 1128.96

$(60.6)^{2}$ = 3672.36

$(-49.4)^{2}$ = 2440.36

$(-18.4)^{2}$ =  338.56

$(- 0.4)^{2}$ =    0.16

$(19.6)^{2}$ =  384.16
 
 
 4. The variance is the average number of these squared differences:

In [70]:
(2061.16+1128.96+3672.36+2440.36+338.56+0.16+384.16) / 7

1432.2457142857143

##### Summary of Calculating the variance
1. Find the mean
2. For each value, find the difference from the mean
3. For each difference, find the square value:
4. The variance is the average number of these squared differences:

Luckily, NumPy has a method to calculate the variance:

In [54]:
z = np.var(speed)
print(z)

1432.2448979591834


As we have learned, the formula to find the standard deviation is the square root of the variance:

var = √1432.25 = 37.85

Or, as in the example from before, use the NumPy to calculate the standard deviation:

In [52]:
speed = [32,111,138,28,59,77,97]

x = numpy.std(speed)

print(x)

37.84501153334721


#### Symbols
Standard Deviation is often represented by the symbol Sigma: $\sigma$

Variance is often represented by the symbol Sigma Square: $\sigma^{2}$



In [71]:
# N/B: just a side note for me on markdown language:

# Latex subscript:
# $x_{2}$

# Latex superscript:
# $x^{2}$

# Latex for the sigma symbol:
# $\sigma$

# Latex for the sigma symbol with the raise to power symbol:
#$\sigma^{2}$

# What are Percentiles?
Percentiles are used in statistics to give you **a number that describes the value that a given percent of the values are lower than.**

Example: Let's say we have an array of the ages of all the people that lives in a street.

ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]

What is the 75. percentile? The answer is 43, meaning that 75% of the people are 43 or younger. We use the NumPy percentile() method to find the percentiles:

The NumPy module has a method for finding the specified percentile:

In [72]:
ages = [5,31,43,48,50,41,7,11,15,39,80,82,32,2,8,6,25,36,27,61,31]

x = numpy.percentile(ages, 75)

print(x)

43.0


In [73]:
# What is the age that 90% of the people are younger than?

y = np.percentile (ages, 90)

print(y)

61.0
